tests/baselines/reference/computedTypesKeyofNoIndexSignatureType.types

```diff
@@ -0,0 +1,43 @@
+=== tests/cases/compiler/computedTypesKeyofNoIndexSignatureType.ts ===
+type Compute<A> = { [K in keyof A]: Compute<A[K]>; } & {};
+>Compute : { [K in keyof A]: { [K in keyof A[K]]: { [K in keyof A[K][K]]: { [K in keyof A[K][K][K]]: { [K in keyof A[K][K][K][K]]: { [K in keyof A[K][K][K][K][K]]: { [K in keyof A[K][K][K][K][K][K]]: { [K in keyof A[K][K][K][K][K][K][K]]: { [K in keyof A[K][K][K][K][K][K][K][K]]: { [K in keyof A[K][K][K][K][K][K][K][K][K]]: { [K in keyof A[K][K][K][K][K][K][K][K][K][K]]: any; }; }; }; }; }; }; }; }; }; }; }
+
+type EqualsTest<T> = <A>() => A extends T ? 1 : 0;
+>EqualsTest : EqualsTest<T>
+
+type Equals<A1, A2> = EqualsTest<A2> extends EqualsTest<A1> ? 1 : 0;
+>Equals : Equals<A1, A2>
+
+type Filter<K, I> = Equals<K, I> extends 1 ? never : K;
+>Filter : Filter<K, I>
+
+type OmitIndex<T, I extends string | number> = {
+>OmitIndex : OmitIndex<T, I>
+
+  [K in keyof T as Filter<K, I>]: T[K];
+};
+
+type IndexObject = { [key: string]: unknown; };
+>IndexObject : IndexObject
+>key : string
+
+type FooBar = { foo: "hello"; bar: "world"; };
+>FooBar : FooBar
+>foo : "hello"
+>bar : "world"
+
+type WithIndex = Compute<FooBar & IndexObject>;   // { [x: string]: {}; foo: "hello"; bar: "world"; } <-- OK
+>WithIndex : { [x: string]: {}; foo: "hello"; bar: "world"; }
+
+type WithoutIndex = OmitIndex<WithIndex, string>; // { foo: "hello"; bar: "world"; }                  <-- OK
+>WithoutIndex : OmitIndex<{ [x: string]: {}; foo: "hello"; bar: "world"; }, string>
+
+type FooBarKey = keyof FooBar;             // "foo" | "bar"   <-- OK
+>FooBarKey : "foo" | "bar"
+
+type WithIndexKey = keyof WithIndex;       // string | number <-- Expected: string 
+>WithIndexKey : string | number
+
+type WithoutIndexKey = keyof WithoutIndex; // number          <-- Expected: "foo" | "bar"
+>WithoutIndexKey : number | "foo" | "bar"

```

Comment:
Is this right? `WithoutIndex` appears, at least in quick info, not to have an index signature at all (not sure if it’s represented differently under the hood), so it seems like you shouldn’t be allowed to index into it with `number`?

Reply:
Slight simplification of the test case, FWIW:

```ts
type OmitIndex<T> = {
    [K in keyof T as (string extends K ? never : K)]: T[K];
};

type WithIndex = { [x: string]: unknown, foo: "hello", bar: "world" };
type WithoutIndex = OmitIndex<WithIndex>; // { foo: "hello"; bar: "world"; }                  <-- OK

type WithIndexKey = keyof WithIndex;
type WithoutIndexKey = keyof WithoutIndex; // number          <-- Expected: "foo" | "bar" 
```

Reply:
_Arguable_ - the `OmitIndex` type is only omitting the `string` key, _however_, a `string` index signature adds both a `string` and `number` key to the `keyof` type, since a `string` index signature implies and allows indexing by `number`!

Reply:
That makes sense, but two follow up questions:

1. Why doesn’t quick info show the number index signature in `WithoutIndex`?
2. How should the type `{ [K in keyof { [key: string]: any }]: K }` be understood? Why isn’t it `{ [x: string]: string | number }` or `{ [x: string]: string & number }` or `{ [x: string]: string, [x: number]: number }`?

Reply:
For 1: The _keyof_ logic includes `number` in the key type, however mapped type mappings iterate over the set of properties/indexes in the original `modifiersType` directly if available (and homomorphic), leading to the discrepancy.
As for 2: We have no principled answer and our behaviors are arbitrarily chosen. I think I went over the inconsistency at a design meeting just after key mappings were introduced and we decided to stick with all the inconsistencies so we didn't break anyone.

Reply:
It feels troubling that

```ts
declare let withoutIndex: WithoutIndex;
withoutIndex[3];
```

is an error even though `keyof typeof withoutIndex` includes `number`. Is there any way to observe the numberyness of the keys besides `keyof`?

Reply:
```ts
type WithIndex = { [x: string]: unknown, foo: "hello", bar: "world" };

type StringOrNumber = keyof WithIndex;
```
is already how keyof already works, so, we allow
```ts
declare const val: WithIndex;

val[12];

val[null as any as number];
```

The `number`-yness of string indexers is pretty visible everywhere it can be shown, IMO. I can't think of somewhere we actually forbid using a number on a string index signature... such a location would probably constitute a bug.

Reply:
> is an error even though keyof typeof withoutIndex includes number. 

You can blame this on how we construct mapped types. While the `keyof` result excludes string, mapped type construction actually doesn't use `keyof` at all for homomorphic mapped types (heh), and instead iterates over the slots visible on the modifiers type. If the `string` indexer has been filtered out it doesn't check "oh, well, maybe a `number` indexer is still appropriate given the filter applied to the modifier", it just drops it.

In any case, `number` is present in the keys is true both before and after this change, so I'm thinking probably doesn't have too much bearing on this change as-is?